From a83463e852f0aa9510844d5f9006de9b5576802d Mon Sep 17 00:00:00 2001
From: Varun Sethi <Varun.Sethi@freescale.com>
Date: Wed, 8 Apr 2015 22:59:03 +0530
Subject: [PATCH 052/124] Add support for IOMMU_MMIO flag.

This flag should be used for setting up access to device memory.

Signed-off-by: Varun Sethi <Varun.Sethi@freescale.com>
Change-Id: Ifc8947c274a260c101a37a29a1795231e9a578ab
Reviewed-on: http://git.am.freescale.net:8181/34634
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Stuart Yoder <stuart.yoder@freescale.com>
---
 drivers/iommu/io-pgtable-arm.c | 5 +++++
 include/linux/iommu.h          | 1 +
 2 files changed, 6 insertions(+)

diff --git a/drivers/iommu/io-pgtable-arm.c b/drivers/iommu/io-pgtable-arm.c
index 937832c..7fa0a27 100644
--- a/drivers/iommu/io-pgtable-arm.c
+++ b/drivers/iommu/io-pgtable-arm.c
@@ -301,6 +301,9 @@ static arm_lpae_iopte arm_lpae_prot_to_pte(struct arm_lpae_io_pgtable *data,
 		if (prot & IOMMU_CACHE)
 			pte |= (ARM_LPAE_MAIR_ATTR_IDX_CACHE
 				<< ARM_LPAE_PTE_ATTRINDX_SHIFT);
+		else if (prot & IOMMU_MMIO)
+			pte |= (ARM_LPAE_MAIR_ATTR_IDX_DEV
+				<< ARM_LPAE_PTE_ATTRINDX_SHIFT);
 	} else {
 		pte = ARM_LPAE_PTE_HAP_FAULT;
 		if (prot & IOMMU_READ)
@@ -309,6 +312,8 @@ static arm_lpae_iopte arm_lpae_prot_to_pte(struct arm_lpae_io_pgtable *data,
 			pte |= ARM_LPAE_PTE_HAP_WRITE;
 		if (prot & IOMMU_CACHE)
 			pte |= ARM_LPAE_PTE_MEMATTR_OIWB;
+		else if (prot & IOMMU_MMIO)
+			pte |= ARM_LPAE_PTE_MEMATTR_DEV;
 		else
 			pte |= ARM_LPAE_PTE_MEMATTR_NC;
 	}
diff --git a/include/linux/iommu.h b/include/linux/iommu.h
index 0546b87..cce2bc9 100644
--- a/include/linux/iommu.h
+++ b/include/linux/iommu.h
@@ -30,6 +30,7 @@
 #define IOMMU_WRITE	(1 << 1)
 #define IOMMU_CACHE	(1 << 2) /* DMA cache coherency */
 #define IOMMU_NOEXEC	(1 << 3)
+#define IOMMU_MMIO	(1 << 4) /* Device memory access */
 
 struct iommu_ops;
 struct iommu_group;
-- 
2.1.0.27.g96db324

